dashboard/app/controllers/api/v1/sections_controller.rb

```diff
@@ -30,38 +30,18 @@ def show
   # POST /api/v1/sections
   # Create a new section
   def create
-    authorize! :create, Section
-
-    # TODO: Push validation into model when old API is fully deprecated
-    # Once this has been done, endpoint can use CanCan load_and_authorize_resource
-    # rather than manually authorizing (above)
-    return head :bad_request unless Section.valid_login_type? params[:login_type]
-
-    valid_script = params[:script] && Script.valid_script_id?(current_user, params[:script][:id])
-    script_to_assign = valid_script && Script.get_from_cache(params[:script][:id])
-
-    section = Section.create(
-      {
-        user_id: current_user.id,
-        name: !params[:name].to_s.empty? ? params[:name].to_s : 'New Section',
-        login_type: params[:login_type],
-        grade: Section.valid_grade?(params[:grade].to_s) ? params[:grade].to_s : nil,
-        script_id: script_to_assign ? script_to_assign.id : params[:script_id],
-        course_id: params[:course_id] && Course.valid_course_id?(params[:course_id]) ?
-          params[:course_id].to_i : nil,
-        code: CodeGeneration.random_unique_code(length: 6),
-        stage_extras: params[:stage_extras] || false,
-        pairing_allowed: params[:pairing_allowed].nil? ? true : params[:pairing_allowed]
-      }
-    )
-    render head :bad_request unless section
+    return head :bad_request unless Section.valid_login_type?(@section.login_type)
 
-    # TODO: Move to an after_create step on Section model when old API is fully deprecated
-    if script_to_assign
-      current_user.assign_script script_to_assign
-    end
+    @section.name = 'New Section' if @section.name.nil_or_empty?
+    @section.user = current_user
+    @section.course_id = nil unless Course.valid_course_id?(@section.course_id)
+    @section.script_id = nil unless Script.valid_script_id?(@section.user, @section.script_id)
```

Comment:
looking for feedback on this: i had lines 37 and 38 in the `before_create` hook in the section model, but this ended up being an enormous change, although mostly in unit tests. basically, every time we call `create :section`, i had to stub out the responses from `Course.valid_course_id?` and `Script.valid_script_id?` or the section's course/script would be set to `nil`, which also feels problematic as it's an unexpected side-effect.

alternatively, any/all of this logic could move to a method in the section model that is just a `validate_create_fields` method rather than leveraging rails' `before_create` and causing confusion.

Reply:
continuing discussion from https://github.com/code-dot-org/code-dot-org/pull/22902 : are lines 37-38 necessary, or will rails validation reject a bad course/script id without them? I may be missing something here, but it seems like you could remove these two lines and see if the tests you linked to in that PR are still passing.

Reply:
i'm realizing now that the tests i pointed you to in that PR that cover invalid courses/scripts was for the _other_ sections controller (the one in `controllers/api/v1/sections_controller.rb`, not `controllers/sections_controller.rb`), so it looks like i need to go back and update the tests after all. sorry for the confusion 😞 

but, back on topic: if i remove lines 37 and 38, an invalid script id responds with `:forbidden` and an invalid course id saves the section with the invalid course associated, so i think we do still need this check

Reply:
@madelynkasula how / where does a bad script id cause a `forbidden` response? My guess would be that happens in the `load_resource` callback before the action code, so this actually would have no effect on that.

As far as not saving a section with an invalid course association, that check should go in a model validation causing the `Section` to be invalid.

As far as not affecting existing tests, how about providing useful (valid) defaults in the [factory](https://github.com/code-dot-org/code-dot-org/blob/a9b35040af512661dd7f1c6181aa70317b5dd05b/dashboard/test/factories/factories.rb#L304-L310). 

(side-note: I just noticed, if you're playing w the factory, line [309](https://github.com/code-dot-org/code-dot-org/blob/a9b35040af512661dd7f1c6181aa70317b5dd05b/dashboard/test/factories/factories.rb#L309-L309) is unnecessary)

Or if the default `section` factory should have no course and script, then add traits:
```ruby
  factory :section do
    sequence(:name) {|n| "Section #{n}"}
    user {create :teacher}
    login_type 'email'

    trait :with_script do
      association :script
    end

    trait :with_course do
      association :course
    end
  end
```

Then it can be called like this
```ruby
    section_without_script_and_course = create :section
    section_with_script_and_course = create :section, :with_script, :with_course
```